app/Policies/LeaveArtifactPolicy.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+namespace App\Policies;
+
+use App\Constants\Permissions;
+use App\LeaveArtifact;
+use App\User;
+use App\Leave;
+use App\Library\UserService;
+use Illuminate\Auth\Access\HandlesAuthorization;
+use Illuminate\Auth\Access\Response;
+
+class LeaveArtifactPolicy {
+
+    use HandlesAuthorization;
+
+    protected UserService $userService;
+
+    public function __construct() {
+        $this->userService = new UserService();
+    }
+
+    /**
+     * Determine whether the user can view any models.
+     */
+    public function viewAny(User $user, Leave $leave): bool {
+        if ($leave->user_id === $user->id) {
+            return true;
+        }
+
+        // users with explicit `view leaves` permission can see all leaves
+        if ($user->can(Permissions::VIEW_ANY_LEAVES)) {
+            return true;
+        }
+
+        // a group manager should be able to view the leaves
+        // of any member of their group
+        if ($user->managesAnyGroupWithMember($leave->user)) {
+            return true;
+        }
+
+        if ($this->userService->doesUserManageAnyGroupWithInstructor($user, $leave->user)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Determine whether the user can view the model.
+     */
+    public function view(User $currentUser, LeaveArtifact $leaveArtifact): bool {
+        // users can see their own leaves
+        if ($currentUser->id === $leaveArtifact->leave->user_id) {
+            return true;
+        }
+
+        // users with explicit `view leaves` permission can see all leaves
+        if ($currentUser->can(Permissions::VIEW_ANY_LEAVES)) {
+            return true;
+        }
+
+        // a group manager should be able to view the leaves
+        // of any member of their group
+        if ($currentUser->managesAnyGroupWithMember($leaveArtifact->leave->user)) {
```

Comment:
wanna move these checks to a method (can just be a private controller method) so the two methods here call the same logic?

Reply:
or all the methods I guess, I hadn't scrolled down